### ISSUE TITLE
resource/aws_ssm_parameter: Remove `overwrite` parameter

### ISSUE THREAD

**Issue by gdavison** [hashicorp/terraform-provider-aws#25636]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

The AWS API uses a PUT operation to create or modify SSM Parameters. By default it will return an error on modify, but setting `Overwrite = true` allows the modification.

The `aws_ssm_parameter` resource exposes this as the parameter `overwrite`. This is redundant for Update operations, since the provider always sets it to `true`. When creating an `aws_ssm_parameter`, setting `overwrite = true` allows the Create operation to act as an Import. This does not match the standard Terraform resource model, so should be removed.

This will be a breaking change.

**Comment by mjburling** [comment]
> When creating an aws_ssm_parameter, setting overwrite = true allows the Create operation to act as an Import. This does not match the standard Terraform resource model, so should be removed.

Hi, @gdavison. I agree that it doesn't match the Terraform resource model, but I'm struggling with this a bit: should the provider be more faithful to the API or more faithful to the Terraform resource model?

I'm currently building something where I was expecting [AWS SSM Parameter Store's `Overwrite` Request Parameter](https://docs.aws.amazon.com/systems-manager/latest/APIReference/API_PutParameter.html#systemsmanager-PutParameter-request-Overwrite) to be exposed in terraform. I have a subset of parameters that users find easier to work with if they use some of Parameter Store's built-in features: the TUI (`awscli`) or GUI (AWS Console) frontends. Developers can quickly adjust parameters and even add new ones on the fly to be consumed in their ephemeral environments, and they aren't forced to go back to update terraform/yaml, commit their changes, and wait for the CI process. However, when that automation is run (for further updates _or_ destruction) terraform is supportive and captures their changes in tfstate. The process is lightweight and helps us keep the hierarchies tidy–no dangling hierarchies upon destruction.

It's more strictly (_statefully_) controlled for production environments. I'll admit, this is an edge case and maybe I'm doing it wrong. I'm torn.

**Comment by orlovmyk** [comment]
And how do I need to handle situations where I want to create new version by Terraform rather than recreated parameter each time?

**Comment by orlovmyk** [comment]
@gdavison Any comment on this? This change is breaking and you haven't give any alternative...?

**Comment by orlovmyk** [comment]
@gdavison From terraform docs
https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/ssm_parameter.html#overwrite

[overwrite](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/ssm_parameter.html#overwrite) - (Optional, Deprecated) Overwrite an existing parameter. If not specified, will default to false if the resource has not been created by terraform to avoid overwrite of existing resource and will default to true otherwise (terraform lifecycle rules should then be used to manage the update behavior).

Can you give me an example how to use **lifecycle** meta-argument to achieve same behaviour? I tried almost everything, but have no clue what you meant by that note.

**Comment by briceburg** [comment]
I too am in the same boat as @orlovmyk  ^^^ and would benefit to see an example of lifecycle rules in the documentation. 

For context: 
We manage ssm parameters in terraform, and previously had the `overwrite = true` attribute set. This (I thought) allowed the value to change. The parameters were never imported and always managed by terraform.

In trying to move to the 5.0 version of AWS provider,  I removed the `overwrite` attribute as it's marked deprecated and am now getting ParameterAlreadyExists errors.

```
│ Error: updating SSM Parameter (/discovery/v1/namespaces/...): ParameterAlreadyExists: The parameter already exists. To overwrite this value, set the overwrite option in the request to true.
```

The terraform looks something like:

```terraform

locals {
  discovery_path = "/discovery/v1/namespaces/${var.config.id}"
  discovery_params = {
    account_name         = "foo"
    account_tier         = "bar"
    ...
  }
}

resource "aws_ssm_parameter" "discovery" {
  for_each = local.discovery_params
  name     = "${local.discovery_path}/${each.key}"
  type     = "String"
  value    = jsonencode(each.value)

  tags = {
    Namespace = var.config.id
  }
}
```

Any ideas on managing these paramters whose values perodically change? /cc @jar-b 

Thanks!

**Comment by briceburg** [comment]
To followup on ^^^, following the migration guide for [aws_ssm_parameter](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/guides/version-5-upgrade#resourceaws_ssm_parameter) I tried importing these resources and am getting "Resource already managed by terraform" error... 

```
module.namespace.aws_ssm_parameter.discovery["lb"]: Importing from ID "/discovery/v1/namespaces/pe-app-dev/lb"...
module.namespace.aws_ssm_parameter.discovery["lb"]: Import prepared!
  Prepared aws_ssm_parameter for import
╷
│ Error: Resource already managed by Terraform
│
│ Terraform is already managing a remote object for
│ module.namespace.aws_ssm_parameter.discovery["lb"]. To import to this address
│ you must first remove the existing object from the state.
```

so kind of at a loss of what to do. We have a many params across many configurations and the removing from state <-> re-importing is not a trivial thing. 

to sumarize, 
* we previously managed SSM parameters with v4 of the provider and `overwrite = true` set.
* upgrading to v5 of the provider sends a deprecation notice, so we remove `overwrite = true`.
  * removing `overwrite = true` results in `ParameterAlreadyExists` errors when applying the configuration, and suggests **adding overwrite = true** which is deprecated?
  * importing the resource results in "Resource already managed by Terraform" error and is tedious

**Comment by ChrisMcKee** [comment]
Not sure why this was marked as depreciated before feeling it out.
The parameter is useful, by all means feel free to have the plan output warn that the keys already exist but removing it when it's available in the api seems backwards.

**Comment by phyber** [comment]
I believe this will break at least one workflow in an environment I work with.

There is a base set of Terraform that creates initial resources in an account, including setting some SSM Parameters. This Terraform uses lifecycling to ignore changes to the SSM Parameter value that it sets.

Later, in another set of Terraform managing other resources for the same account, these SSM Parameters are referenced via a `data` lookup.

Some resources are created (in Kubernetes) based on that SSM Parameter before the SSM Parameter is finally overwritten with a new value. That resource in Kubernetes does its thing and then ceases to exist. We do not want Terraform to create the resource on every `terraform apply`, so it's controlled via an SSM Parameter.

Because this manipulation of the same SSM Parameter happens in different Terraform workspaces (and separate state files), I believe that the overwriting of the SSM Parameter that we do here would cause an error if the `overwrite` parameter is removed.

The parameter wouldn't exist in the 2nd state file, which I believe means that Terraform wouldn't know to set the `overwrite` parameter to `true` internally, and the API would throw an error since the SSM Parameter already exists in AWS.

**Comment by briceburg** [comment]
@jar-b thanks for the update and clear explanation.  is there a workaround for 5.0 currently to avoid the deprecation notice? in my situation I am;
* unable to import the resource (`Resource already managed by Terraform` error). 
* unable to update the resource without setting overwrite to true (`ParameterAlreadyExists`)

thus I have to set overwrite to true and see the deprecation notice. I think my use-case is that I always have managed the ssm parameter via the terraform configuration (in v4 of the aws provider), but  decided to set `overwrite` to true  (it wasn't deprecated) in case the parameter got updated outside of terraform and thus terrafrom would revert the external change (kind of like crossplane).

**Comment by jar-b** [comment]
Hey @briceburg - Based on your description is sounds like you're expecting the standard Terraform workflow, mentioned above here:

> Lastly, and unfortunately, configurations expecting the standard update flow will need to keep `overwrite = true` set until this becomes the default behavior in` v6.0.0`. Removing it in `v5.X` will result in the default value of `false`, preventing the parameter value from being updated.

There isn't currently a workaround to avoid the deprecation notice in `v5.X`, but when this becomes the default behavior in `v6.0.0` you can simply remove the `overwrite` parameter and retain the same behavior.

**Comment by briceburg** [comment]
Ok great! We can live with it =) I wanted to doubly make sure our terraforms were not deviating _too_ far towards exotic... 

So in a post v6+ world, where we have an SSM parameter created by an "upstream" terraform configuration and potentially updated by a "downstream" one, could we use the following pattern? 

## upstream terraform

```terraform
# create addon discovery params
resource "aws_ssm_parameter" "addons" {
  for_each = toset(["foo", "bar"])
  name     = "/disovery/v1/addons/${each.key}"
  type     = "String"
  value    = jsonencode({ enabled : false })

  # ignore subsequent 'downstream' changes
  lifecycle {
    ignore_changes = [value, tags]
  }
}
```

## downstream terraform
```terraform

import {
  id = "/disovery/v1/addons/foo"
  to = aws_ssm_parameter.discovery
}

# enable the 'foo' addon
resource "aws_ssm_parameter" "discovery" {
  name = "/disovery/v1/addons/foo"
  type = "String"
  value = jsonencode({
    enabled = true
    hello   = "world"
  })
} 
```

**Comment by jar-b** [comment]
Yes, the `import` block would be the new pattern to achieve the "import-on-create" behavior `overwrite = true` previously enabled for the "downstream" resource.

The [import docs](https://developer.hashicorp.com/terraform/language/import) also mention leaving the block permanently as a record of the resource origin, so I don't think this approach deviates from the intended of the feature.
> The import block records that Terraform imported the resource and did not create it. After importing, you can optionally remove import blocks from your configuration or leave them as a record of the resource's origin.

**Comment by sirocode** [comment]
Hi, the overwrite feature has been working for years and it's a crucial part for a tons of deployments in my company now. We store docker tags (as well as other deployment-wide params) of the images in SSM param store and using "terraform apply -var image_ver=1234" to deploy new image to AWS ECS.

In case when image_ver var is not specified the image version gets collected using terraform data. It's a bullet proof schema that has been working like a charm for years.

**Comment by arnvid** [comment]
In our case we use a variable to set part of the ssm path - and the new import block does not support this.

```
import {
  to = aws_ssm_parameter.bootstrap_complete
  id = "${local.tlz_prefix}/bootstrap_complete"
}

resource "aws_ssm_parameter" "bootstrap_complete" {
  name           = "${local.tlz_prefix}/bootstrap_complete"
  type           = "String"
  insecure_value = var.tlz_bootstrap_complete
}
```

**Comment by dmajano** [comment]
In our case, we have the creation of every ssm parameter store with a "null" value first, and after, in another terraform stack, we define the value of these parameters.

The creation part with `ignore_changes` is covered:

```
resource "aws_ssm_parameter" "app" {
  for_each = toset(local.app_parameters)
  name  = "${local.app_path}/${each.key}"
  type  = "String"
  value = "null"

  lifecycle {
    ignore_changes        = all
  }
}
```
The second part, update the parameter's value without overwrite is very painful because we have all our terraform projects developed with nested modules and the `import` block only can be used at root level:

```
│ Error: Invalid import configuration
│ 
│   on ../../globals/app/main.tf line 5:
│    5: import {
│ 
│ An import block was detected in "module.globals_app". Import blocks are only allowed in the root module.
╵
╷
│ Error: Variables not allowed
│ 
│   on ../../globals/app/main.tf line 7, in import:
│    7:   id = "${local.path}/example"
│ 
│ Variables may not be used here.

```

While with `overwrite = true` everything is completely simplified

**Comment by jbonnier** [comment]
Kind a weird that there is no in the middle workaround.

Usually, when something is deprecated. The future configuration is implemented in a non-breaking way with a deprecation warning. And in the next major release, the deprecated functionality is removed.

If I understood this thread correctly, there is no way to implement the new configuration before moving to v6.0.0 which is kinda weird.

**Comment by pvilasra** [comment]
I am using terraform v1.1.2 version If I want to use import block instead of overrite = "true" , The import blocks are only available in Terraform v1.5.0 and later. If I use terraform import command its says resource already managed by terraform. If I remove overrrite parameter it says resource already created. So I need to upgrade terraform version to latest v1.6.0 for this change ?  or is there any other workaround for this?

**Comment by cmd-aj** [comment]
I have a **module** that uses this argument. How would I modify my code going forward? 

```hcl
# Look up the external ID in AWS SSM Parameter Store if not provided
data "aws_ssm_parameter" "external_id" {
  count = var.cross_account_role_external_id == null ? 1 : 0
  name  = var.cross_account_role_external_id_parameter_store_name
}

# Create the external ID in AWS SSM Parameter Store if provided
resource "aws_ssm_parameter" "external_id" {
  count = var.cross_account_role_external_id != null ? 1 : 0

  name        = var.cross_account_role_external_id_parameter_store_name
  type        = "SecureString"
  value       = var.cross_account_role_external_id
  description = "External ID for cross account role access"
  key_id      = aws_kms_key.this.arn
  overwrite   = true
}```

**Comment by kovaacs** [comment]
I agree that it's very strange to get a deprecation warning without any action possible until the actual breaking change. 

Either there should be
- no warning now and this should be considered a breaking change in 6.0, or 
- warning now with a clear way to accommodate the change coming in 6.0.

What we have now is the worst of both worlds. Correct me if I'm wrong.

**Comment by faust64** [comment]
Hi,

I'm a little confused as well ... I did attempt to remove that "overwrite" from my code, at which point terraform is unable to write values, when rotating some secret.

Can't share exact code here, sorry, ... imagine I have an iam_access_key. its value written as an SSM parameter. And a time_rotating.
My access key is rotating on an interval. When it does, Terraform plan show my SSM param should be re-created. Omitting that "overwrite=true", Terraform apply fails to write new value. Tells me param already exists.

Would make sense for replacement to first delete old value, and only then write new copy.
Would be nice being able to test that code change, before deprecation turns into full removal.

Thanks!

**Comment by queglay** [comment]
I am very confused after coming across this deprecation warning, so I disabled the overwrite value.
but then when I try to import the resource it says I cannot as its already managed by terraform.  and then if I apply I get:
```
Error: updating SSM Parameter (/firehawk/resourcetier/dev/vault_kms_unseal_key_id): ParameterAlreadyExists: The parameter already exists. To overwrite this value, set the overwrite option in the request to true.
```

What should I do to adhere to the deprecation warning and deprecate the parameter while still allowing me to set parameter values?

**Comment by tjstansell** [comment]
In some cases, we use the overwrite flag to ensure that terraform takes over any existing parameter store entries that might have been manually added in the past.  they don't necessarily exist.  if we try to use an import block, it will fail if it doesn't exist.  There's no way to say "import if it exists, otherwise create it", which the current `overwrite=true` attribute gives us.

**Comment by cobbr2** [comment]
We are also in a spot where there is no sensible `import`. Even using terraform 1.8.3, when we try an import like:
```
import {
  id = "/rc/sql/rc/database_readonly_url"
  to = "module.global.module.db_postgres[0].module.readonly_url.aws_ssm_parameter.region01"
}
```
we get lovely fun like:
```
│ Error: Invalid import address
│
│   on import-manual-ssm.tf line 3, in import:
│    3:   to = "module.global.module.db_postgres[0].module.readonly_url.aws_ssm_parameter.region01"
│
│ Only managed resources can be imported.
╵

╷
│ Error: Invalid expression
│
│   on import-manual-ssm.tf line 3, in import:
│    3:   to = "module.global.module.db_postgres[0].module.readonly_url.aws_ssm_parameter.region01"
│ A single static variable reference is required: only attribute access and indexing with constant keys. No calculations, function calls, template expressions, etc are allowed here.
```
AFAICT, that expression is only accessing attributes and indexing with constant keys, so that second error makes no sense.  

At the moment, we're actually having to *add* `overwrite=true` into our modules because things that should be getting imported and set aren't -- and can't be given the above import problems.

**Comment by tjstansell** [comment]
Your import should be:
```
import {
  id = "/rc/sql/rc/database_readonly_url"
  to = module.global.module.db_postgres[0].module.readonly_url.aws_ssm_parameter.region01
}
```
No quotes for the `to` value.

**Comment by edwardofclt** [comment]
Please don't do this; our businesses rely heavily on SSM resource and the overwrite value is what allows us to programmatically update value without manual effort.

**Comment by onprema** [comment]
So, what is the recommended way of updating an ssm parameter that already exists?

**Comment by joaocc** [comment]
@onprema did you think about using the new improved import blocks, which apparently allow even dynamic parameters?

**Comment by onprema** [comment]
> @onprema did you think about using the new improved import blocks, which apparently allow even dynamic parameters?

Thanks for responding, @joaocc! Do you know if that handles cases where the parameter doesn't exist? I assume the import would fail if that were the case.

**Comment by joaocc** [comment]
not sure. we are still in 1.5.x, and newer versions seem to have improved capabilities a lot.

**Comment by faust64** [comment]
How does import block helps here?
We need terraform being able to write some secrets, and rotate them on a schedule. Periodically write a new value. Overwrite it, if you will.

**Comment by tisted** [comment]
The lifecycle option to prevent changes is already there. It seems counter-intuitive to remove a feature that people actively use when you could just include a "here-be-dragons" in the docs.

**Comment by cjwooo** [comment]
```
Lastly, and unfortunately, configurations expecting the standard update flow will need to keep overwrite = true set until this becomes the default behavior in v6.0.0
```

Can someone please confirm if this is still true or not, even after https://github.com/hashicorp/terraform-provider-aws/pull/32372 was released? i.e. Can I safely remove the `overwrite = true` on aws_ssm_parameter resources, and still be able to update the parameter values? The amount of discussion on this topic makes me nervous, and I would like to have this clarified before I go in and update dozens of TF resources.

We are using provider version 5.72.1.

**Comment by breathingdust** [comment]
Closing as this functionality is clearly heavily relied on by the community.

**Comment by jar-b** [comment]
Hey everyone - to add some additional context, in preparing for the V6.0.0 major release we reviewed all of the comments on this issue and weighed the disruption in existing use cases against following Terraform provider developments best practices. Given the number of comments describing large scale deployments of this resource which rely on the existing behavior of the `overwrite` argument, removing it was not an acceptable trade off.

I'll be removing the deprecation argument shortly and configurations relying on the `overwrite` argument can continue doing so into the future. For those which have already migrated to the explicit `import` blocks and/or lifecycle rules, those will also continue to function as intended without changes. Thanks for your patience and engagement on this issue.
